scripts/anvil_tools/set_up_anvil_workspaces.py

```diff
@@ -0,0 +1,320 @@
+    create_ws_request["attributes"] = {}
+    # TODO: set noWorkspaceOwner = True for data delivery workspaces - picard svc is the only owner
+    create_ws_request["noWorkspaceOwner"] = False
+
+    return create_ws_request
+
+
+def add_member_to_auth_domain(auth_domain_name, email, permission):
+    """Add group with given permissions to authorization domain."""
+
+    # request URL for addUserToGroup
+    uri = f"https://api.firecloud.org/api/groups/{auth_domain_name}/{permission}/{email}"
+
+    # Get access token and and add to headers for requests.
+    # -H  "accept: */*" -H  "Authorization: Bearer [token]"
+    headers = {"Authorization": "Bearer " + get_access_token(), "accept": "*/*"}
+
+    # capture response from API and parse out status code
+    response = requests.put(uri, headers=headers)
+    status_code = response.status_code
+
+    if status_code != 204:  # AD update with member fail
+        print(f"WARNING: Failed to update Authorization Domain, {auth_domain_name}, with group: {email}.")
+        print("Check output file for error details.")
+        # return None, response.json()
+        return False, status_code, response.json()
+
+    # AD update with member success
+    print(f"Successfully updated Authorization Domain, {auth_domain_name}, with group: {email}.")
+    return True, status_code, None
+
+
+def setup_auth_domain(auth_domain_name):
+    """Create the Terra goolge group to be used as auth domain on workspaces."""
+
+    # request URL for createGroup
+    uri = f"https://api.firecloud.org/api/groups/{auth_domain_name}"
+
+    # Get access token and and add to headers for requests.
+    # -H  "accept: application/json" -H  "Authorization: Bearer [token]"
+    headers = {"Authorization": "Bearer " + get_access_token(), "accept": "application/json"}
+
+    # capture response from API and parse out status code
+    response = requests.post(uri, headers=headers)
+    status_code = response.status_code
+
+    if status_code in [403, 409]:                    # if AD already exists
+        ad_exists_message = "Authorization Domain with name already exists. Try again with unique group name."
+        print(f"WARNING: Failed to setup Authorization Domain with name: {auth_domain_name}.")
+        return status_code, None, ad_exists_message, None  # 403 or 409
+    if status_code != 201:                           # other error (500)
+        print(f"WARNING: Failed to create Authorization Domain with name: {auth_domain_name}. Check output file for error details.")
+        return status_code, response.json(), None, None
+
+    # if AD creation success - status_code == 201
+    anvil_admin_group_name = "anvil-admins@firecloud.org"
+    is_add_mem, add_mem_status_code, add_mem_response = add_member_to_auth_domain(auth_domain_name, anvil_admin_group_name, "ADMIN")  # 204 - success otherwise fail
+
+    if not is_add_mem:           # add member to AD fail
+        return None, add_mem_response, None, False
```

Comment:
why are you returning no status code for `setup_ad_status_code` here? the group was created, but there was an error adding members, right?

Reply:
oh I think I see - the first thing returned is the success of the entire process, which failed, and you're indicating the reason for failure with the second thing returned. I think this works! It would help a lot to explicitly say what these four values represent (see previous comment)

Reply:
Oh that was because I'm using the boolean value to determine if i should print the `add_mem_response` and don't use the status_code. Should I be returning it anyway?